packable/packable/src/packable/set.rs

```diff
@@ -0,0 +1,103 @@
+pub enum UnpackSetError<T, I, P> {
+    /// A duplicate set item.
+    DuplicateItem(T),
+    /// Semantic error raised while unpacking an item of the sequence. Typically this is
+    /// [`Packable::UnpackError`](crate::Packable::UnpackError).
+    Item(I),
+    /// Semantic error raised when the length prefix cannot be unpacked.
+    Prefix(P),
+}
+
+impl<T, I: fmt::Debug, P: fmt::Debug> fmt::Debug for UnpackSetError<T, I, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::DuplicateItem(_) => f.debug_tuple("DuplicateItem").finish(),
+            Self::Item(arg0) => f.debug_tuple("Item").field(arg0).finish(),
+            Self::Prefix(arg0) => f.debug_tuple("Prefix").field(arg0).finish(),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T, I, P> std::error::Error for UnpackSetError<T, I, P>
+where
+    I: std::error::Error,
+    P: std::error::Error,
+{
+}
+
+impl<T, I, P> From<Infallible> for UnpackSetError<T, I, P> {
+    fn from(err: Infallible) -> Self {
+        match err {}
+    }
+}
+
+impl<T, I: fmt::Display, P: fmt::Display> fmt::Display for UnpackSetError<T, I, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::DuplicateItem(_) => write!(f, "duplicate item in set"),
+            Self::Item(err) => write!(f, "cannot unpack item: {}", err),
+            Self::Prefix(err) => write!(f, "cannot unpack prefix: {}", err),
+        }
+    }
+}
+
+#[cfg(feature = "usize")]
+mod btreeset {
+    use alloc::collections::BTreeSet;
+
+    use super::*;
+    use crate::{error::UnpackError, packer::Packer, unpacker::Unpacker, Packable};
+
+    impl<T: Packable + Ord> Packable for BTreeSet<T> {
+        type UnpackError = UnpackSetError<T, T::UnpackError, <usize as Packable>::UnpackError>;
+        type UnpackVisitor = T::UnpackVisitor;
+
+        #[inline]
+        fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+            // This cast is fine because we know `usize` is not larger than `64` bits.
+            (self.len() as u64).pack(packer)?;
+
+            for item in self.iter() {
+                item.pack(packer)?;
+            }
+
+            Ok(())
+        }
+
+        #[inline]
+        fn unpack<U: Unpacker, const VERIFY: bool>(
+            unpacker: &mut U,
+            visitor: &Self::UnpackVisitor,
+        ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+            use crate::error::UnpackErrorExt;
+
+            let len = u64::unpack::<_, VERIFY>(unpacker, &())
+                .coerce()?
+                .try_into()
+                .map_err(|err| UnpackError::Packable(Self::UnpackError::Prefix(err)))?;
+
+            let mut set = BTreeSet::new();
+
+            for _ in 0..len {
+                let item = T::unpack::<_, VERIFY>(unpacker, visitor).map_packable_err(Self::UnpackError::Item)?;
+                if set.contains(&item) {
```

Comment:
If you are not using `insert`'s return just to keep the item for the error, tbh I'm not sure that's worth it, I'd remove the item from the error

Reply:
You don't even print it in Debug/Display...

Reply:
You can use insert's return type to see if it was contained or not

Reply:
Yeah but the item in the error is not very helpful tbh. If you really think it's helpful then log it in Debug. But contains/insert for each element of the set sounds like a lot for not much

Reply:
But the debug bounds propagate to the packable impl :c I think it's good to have the item in the value for usages where it can be handled in code

Reply:
I'll leave it up to your judgment, it sounds pretty unfriendly to get access to this item, we'd have to match on a nested error. And it degrades the performances.

Reply:
I don't think it degrades performance at all 😆

Reply:
I don't see why it wouldn't degrade performance, that's an additional hash that we could remove. But as said, I'll leave it up to you.